packages/cloud_firestore/cloud_firestore_platform_interface/lib/cloud_firestore_platform_interface.dart

```diff
@@ -0,0 +1,236 @@
+/// changes. Extending this class (using `extends`) ensures that the subclass
+/// will get the default implementation, while platform implementations that
+/// `implements` this interface will be broken by newly added
+/// [CloudFirestorePlatform] methods.
+abstract class CloudFirestorePlatform {
+  /// Only mock implementations should set this to `true`.
+  ///
+  /// Mockito mocks implement this class with `implements` which is forbidden
+  /// (see class docs). This property provides a backdoor for mocks to skip the
+  /// verification that the class isn't implemented with `implements`.
+  @visibleForTesting
+  bool get isMock => false;
+
+  /// The default instance of [CloudFirestorePlatform] to use.
+  ///
+  /// Platform-specific plugins should override this with their own class
+  /// that extends [CloudFirestorePlatform] when they register themselves.
+  ///
+  /// Defaults to [MethodChannelCloudFirestore].
+  static CloudFirestorePlatform get instance => _instance;
+
+  static CloudFirestorePlatform _instance = MethodChannelCloudFirestore();
+
+  // TODO(amirh): Extract common platform interface logic.
+  // https://github.com/flutter/flutter/issues/43368
+  static set instance(CloudFirestorePlatform instance) {
+    if (!instance.isMock) {
+      try {
+        instance._verifyProvidesDefaultImplementations();
+      } on NoSuchMethodError catch (_) {
+        throw AssertionError(
+            'Platform interfaces must not be implemented with `implements`');
+      }
+    }
+    _instance = instance;
+  }
+
+  /// This method ensures that [CloudFirestorePlatform] isn't implemented with `implements`.
+  ///
+  /// See class docs for more details on why using `implements` to implement
+  /// [CloudFirestorePlatform] is forbidden.
+  ///
+  /// This private method is called by the [instance] setter, which should fail
+  /// if the provided instance is a class implemented with `implements`.
+  void _verifyProvidesDefaultImplementations() {}
+
+  // Actual API
+  // Global
+  /// Removes any listener by its handle.
+  /// All handles must be unique across al types of listeners.
+  Future<void> removeListener(int handle) async {
+    throw UnimplementedError('removeListener() is not implemented');
+  }
+
+  // Firestore
+  Future<void> enablePersistence(String app, {@required bool enable}) async {
+    throw UnimplementedError('enablePersistence() is not implemented');
+  }
+
+  Future<void> settings(
+    String app, {
+    bool persistenceEnabled,
+    String host,
+    bool sslEnabled,
+    int cacheSizeBytes,
+  }) async {
+    throw UnimplementedError('settings() is not implemented');
+  }
+
+  Future<Map<String, dynamic>> runTransaction(
+    String app, {
+    @required PlatformTransactionHandler transactionHandler,
+    int transactionTimeout,
+  }) async {
+    throw UnimplementedError('runTransaction() is not implemented');
+  }
+
+  // Document Reference
+  Future<void> setDocumentReferenceData(
```

Comment:
@amirh Does this seem right to you as a style convention for federated plugins that are wrapping instances of native classes other than the main class of the plugin?

Right now we're shoehorning everything onto `CloudFirestorePlatform` instead of having a `PlatformDocumentReference` that has the same extends-only semantics as `CloudFirestorePlatform`.

If we wanted to be less idiomatically Dart but make the class/method name more clear we could call it something like `DocumentReference_setData`

Reply:
I think it's definitely makes sense for the platform interface to be composed from multiple classes. In fact we've already done this in the webview plugin (WebViewPlatform and WebViewPlatformController). I think it's less of a convention thing and more of a "whatever design works fits better for a given plugin".

For this case, after discussing with Collin, it seems like multiple classes is a better choice.